Add some special cases for universal compare

dolio opened the pull request:
Handle `Float` and `Nat` specifically, since comparing them as if they were `Int` doesn't work correctly.

Addresses #2257

dolio committed:
Add some special cases for universal compare

- Handle `Float` and `Nat` specifically, since comparing them as if they
  were `Int` doesn't work correctly.

dolio committed:
Add tests for compare corner cases

pchiusano commented on parser-typechecker/src/Unison/Runtime/Stack.hs:
```diff
@@ -193,6 +193,21 @@ universalEq frn = eqc False
+  ba = byteArrayFromList [i,j]
+  ni, nj :: Word
+  ni = indexByteArray ba 0
+  nj = indexByteArray ba 1
```
Is there no more direct way to do this, like casting from `Int` to `Float` / `Nat` ?

dolio replied:
Hmm, I think I can do better. `DataC` isn't the most direct way to access these things.

Just casting `Int` to `Double` is not guaranteed to work, though.

dolio replied:
Oh, but actually, the way these are going to be represented in practice is no better. We'll get `Int` values, because they will be specifically built with constructors for 1 unboxed value.

pchiusano replied:
I mean the whole byte array business. Why not just use:

https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Float.html#v:int2Double 

To go from `Int` to `Double` safely. And then not sure how to go from `Int -> Word` bitwise... will `fromIntegral` do it? Or does that truncate at 0?

dolio replied:
`int2Double` is like `1 -> 1.0`. `fromIntegral` should work for words, though, yeah.

pchiusano replied:
Ah, gotcha. Is there really no more direct way to bitwise cast from `Int <-> Double`? There's gotta be.

I just thought of something else, which is that I want `Universal.compare` to be a total order for floats, not the IEEE partial order thing. If you want the partial order you can always use `Float.<` etc. Basically keep the Universal functions as structural/lexicographical.

I think you just have to flip the sign bit then compare the ints?

pchiusano commented on parser-typechecker/src/Unison/Runtime/Stack.hs:
```diff
@@ -203,6 +218,9 @@ universalCompare frn = cmpc False
     = compare (length l) (length r) <> fold (zipWith cm l r)
+  cmpc _ (DataC rf1 0 [i] []) (DataC rf2 0 [j] [])
+    | rf1 == Ty.floatRef && rf2 == Ty.floatRef = compareAsFloat i j
+    | rf1 == Ty.natRef && rf2 == Ty.natRef = compareAsNat i j
```
Bummer that we're having to compare full `Reference` values here, rather than a int tag. It slows all other comparisons down and when there's a hit it has to look at all 512 bits of the hash to prove that. 

Do we not currently have an integer tag for the type of boxed values?

pchiusano replied:
One thought I had if we're missing an int tag for boxed types - just comparing the first K bytes of the hash is sufficient to establish equality for this purpose. Like K = 10 is  2^80 = 1,208,925,819,614,629,174,706,176... how many types would we need to hit a 1% chance of collision?

dolio replied:
I don't really understand what you're asking. The reference checks already only occur when both arguments have exactly 1 unboxed field, and have tag 0. So, only stuff like `Int` `Nat` and `Float`, basically, if that's what you're worried about.

pchiusano replied:
Ah, yeah, I was thinking it slowed everything down. Also ignore what I said about hashes, since these are builtin types.

When that pattern is matched, though it will end up doing a text comparison of the text `##Nat` or `##Float` (because `Ty.natRef` is `Builtin "##Nat"`) for every universal comparison of those kinds of boxed values. Which is a bit silly. 

Do you not have an integer tag for the _type_ of the thing? (I don't mean the constructor id tag, which is 0 in this case)

dolio replied:
At some point, I was packing some type information into the data tags. But I think that got eliminated when the entire reference was put in there.

dolio replied:
I can try to add the type information back in. I think maybe the reason I got rid of it was that it was more difficult to handle for serialization. I think it's possible, though. Just a bit tricky.

pchiusano replied:
Okay, cool. I'd say don't sweat it for now. We aren't even sure it's much slower.

Something we could also consider for later is just not having `Text` in the `Builtin` side of a reference - use an enum which we can convert to and from text. This could speed up the comparisons maybe but also give us coverage checking in places where we're inspecting builtins.

dolio committed:
Use `fromIntegral` in universal Nat comparison

dolio committed:
Add reference information back to tags

- Reference numbers are packed into tags on data creation
- When dumping the data to the stack, reference numbers are masked off
- Serialization removes/restores the reference part of the tag for
  portability.
- Universal comparison ops take advantage of the tag to avoid comparing
  references, which could be slow.

dolio committed:
Rework universal comparison for Float
